lib/segment/src/index/field_index/stat_tools.rs

```diff
@@ -0,0 +1,145 @@
+    // Probability to select each unique value
+    let prob_select = 1. - prob_not_select(total_values, values_per_point, selected_values_count);
+    prob_select * total_points as f64
+}
+
+/// Fast approximate computation of $ln(n!)$
+/// See: <https://en.wikipedia.org/wiki/Stirling%27s_approximation>
+fn approx_fact_log(n: f64) -> f64 {
+    if n < 1.0 {
+        return 1.0; // By definition
+    }
+    (2. * PI * n).sqrt().ln() + n * (n / E).ln()
+}
+
+/// Probability of each individual unique point to be selected with the query
+///
+/// Straight equation:
+///     $\prod_{i=0}^{N-1} \frac{total - avg - i}{total - i}$
+/// , where `N` - number of selected points
+///
+/// Prove:
+///
+/// $$
+/// \prod_{i=0}^{N-1} \frac{total - avg - i}{total - i}
+///     = \frac{\prod_{i=0}^{N-1} (total - avg - i)}{\prod_{i=0}^{N-1}(total - i)}
+///     = \frac{\prod_{i=1}^{N} (total - avg - i + 1)}{\prod_{i=1}^{N}(total - i + 1)}\\
+///     = \frac{\prod_{i=1}^{N} (total - avg - (N - i + 1) + 1)}{\prod_{i=1}^{N}(total - (N - i + 1) + 1)}
+///     = \frac{\prod_{i=1}^{N} (i + total - avg - N)}{\prod_{i=1}^{N}(i + total - N)}\\
+///     = \frac{\prod_{i=1}^{total - avg} i}{\prod_{i=1}^{total - avg - N} i} \frac{\prod_{i=1}^{total - N} i}{\prod_{i=1}^{total} i}
+///     = \frac{(total - avg)!(total - N)!}{(total - avg - N)!(total)!}
+///     = \exp(\ln{\frac{(total - avg)!(total - N)!}{(total - avg - N)!(total)!}})\\
+///     = \exp(\ln((total - avg)!(total - N)!) - \ln((total - avg - N)!(total)!))
+///     = \exp( \ln((total - avg)!) + \ln((total - N)!) - \ln((total - avg - N)!) - \ln(total!))
+/// $$
+///
+/// Hint: use <https://latex.codecogs.com/eqneditor/editor.php> to render formula
+fn prob_not_select(total: usize, avg: f64, selected: usize) -> f64 {
+    let total = total as f64;
+    let selected = selected as f64;
+    (approx_fact_log(total - avg) + approx_fact_log(total - selected)
+        - approx_fact_log(total - avg - selected)
+        - approx_fact_log(total))
+    .exp()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand::prelude::StdRng;
+    use rand::seq::SliceRandom;
+    use rand::SeedableRng;
+    use std::collections::HashSet;
+
+    fn simulate(uniq: usize, avg: usize, selected: usize) -> usize {
+        let mut data: Vec<_> = vec![];
+        for i in 0..uniq {
+            for _ in 0..avg {
+                data.push(i);
+            }
+        }
+        data.shuffle(&mut StdRng::seed_from_u64(42));
+
+        let mut unique_selected: HashSet<_> = Default::default();
+        for x in data.into_iter().take(selected) {
+            unique_selected.insert(x);
+        }
+
+        unique_selected.len()
+    }
+
+    #[test]
+    fn approx_factorial() {
+        let approx = approx_fact_log(10.).exp();
+        let real = (2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * 10) as f64;
+        let error = (approx / real - 1.0).abs();
+        assert!(error < 0.01);
+    }
+
+    #[test]
+    fn test_estimation_corner_cases() {
```

Comment:
anything to actually assert in this test case?

Reply:
I ended up allowing some strange behavior is some rare corner cases, so it should be corrected in the further code. Added check for NaN, though